pkg/bpf/map_linux_test.go

```diff
@@ -384,6 +387,97 @@ func (s *BPFPrivilegedTestSuite) TestDump(c *C) {
 	})
 }
 
+func (s *BPFPrivilegedTestSuite) TestDumpReliablyWithCallback(c *C) {
+	maxEntries := uint32(256)
+	m := NewMap("cilium_dump_test",
+		MapTypeHash,
+		&TestKey{},
+		int(unsafe.Sizeof(TestKey{})),
+		&TestValue{},
+		int(unsafe.Sizeof(TestValue{})),
+		int(maxEntries),
+		BPF_F_NO_PREALLOC,
+		0,
+		ConvertKeyValue,
+	).WithCache()
+	_, err := m.OpenOrCreate()
+	if err != nil {
+		c.Assert(err, IsNil)
+	}
+	defer func() {
+		path, _ := m.Path()
+		os.Remove(path)
+	}()
+	defer m.Close()
+
+	for i := uint32(4); i < maxEntries; i++ {
+		err := m.Update(&TestKey{Key: i}, &TestValue{Value: i + 100})
+		c.Check(err, IsNil) // we want to run the deferred calls
+	}
+	// start a goroutine that continuously updates the map
+	started := make(chan struct{}, 1)
+	done := make(chan struct{}, 1)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		started <- struct{}{}
+		for {
+			for i := uint32(0); i < 4; i++ {
+				select {
+				case <-done:
```

Comment:
Why is this needed? The go routine will always complete anyway

Reply:
It will not; the inner `for` loop is within an infinite `for {}` loop.
The idea is that this routine constantly updates/deletes the first 4 keys in the map while the other routine calls `DumpReliablyWithCallback`.

Reply:
I moved the `select` statement to the outer loop for better clarity.